compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java

```diff
@@ -411,6 +411,14 @@ public void visit(BLangWhile whileNode) {
             this.scope = this.symbolEnv;
             this.acceptNode(whileNode.body, this.symbolEnv);
         }
+        visitOnFailClause(whileNode.onFailClause);
+    }
+
+    private void visitOnFailClause(BLangOnFailClause onFailClause) {
+        if (onFailClause != null) {
+            SymbolEnv onFailEnv = SymbolEnv.createOnFailEnv(onFailClause, this.symbolEnv);
+            this.acceptNode(onFailClause, onFailEnv);
+        }
```

Comment:
Shouldn't this logic be in the visit() method for the on fail clause?

Reply:
this was to avoid null check duplication within multiple nodes. we can move the environment creating to onfail visit if needed. wdyt?

Reply:
this was to avoid null check duplication in multiple places

Reply:
we can move the env creation to onfail clause if needed? wdyt?

Reply:
There won't be any duplication right? What I meant was to move this logic to the `visit(BLangOnFailClause)` method. And since the on fail clause seem to have a block stmt, I think we can just simply call `acceptNode(onFail.body, symbolEnv)` in the visit.

Reply:
Updated in 1b130a542273ac48fa116aeee3fb56c14953dba2